### PR TITLE
[otbn] Assertions for SEC_WIPE countermeasure, minor fix

### PR DESCRIPTION
**Pull request by ctopal**
Goes towards completing https://github.com/lowRISC/opentitan/issues/13964. While working on assertions, found a minor bug in the code regarding an extra unused register in the register file, removed that as well.

### CODE REVIEW

File: hw/ip/otbn/rtl/otbn_rf_base_ff.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn_rf_base_ff.sv:51]: I suggest to move this change to a separate PR, since it is orthogonal to secure wipe DV.

File: hw/ip/otbn/rtl/otbn_alu_bignum.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn_alu_bignum.sv:850]: Nit: add spaces in concatenation

File: hw/ip/otbn/rtl/otbn.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn.sv:1202]: Should we also add an assertion that `busy_secure_wipe` rises after every reset and before OTBN starts executing? I know the model has an assertion for this too, but I think this is important enough to be checked also in RTL.

File: hw/ip/otbn/rtl/otbn.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn.sv:1213]: Clever! :slightly_smiling_face:

File: hw/ip/otbn/rtl/otbn.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn.sv:1199]: Nit: add a space between `;` and `i`

File: hw/ip/otbn/rtl/otbn.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn.sv:1218]: Nit: add a space between `;` and `i`

File: hw/ip/otbn/rtl/otbn.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn.sv:1247]: I would change the assertion name to `SecWipeNonZeroMod_A`, since the secure wipe should *not* zero MOD

File: hw/ip/otbn/rtl/otbn.sv
Comment (andreaskurth) [inline hw/ip/otbn/rtl/otbn.sv:1251]: dito change of assertion name

**Review by andreaskurth — COMMENTED** [review]
Thanks @ctopal, this essentially LGTM. I have a few comments/requests

File: hw/ip/otbn/rtl/otbn_rf_base_ff.sv
Comment (ctopal) [reply hw/ip/otbn/rtl/otbn_rf_base_ff.sv:51]: Good suggestion, done! https://github.com/lowRISC/opentitan/pull/14431

File: hw/ip/otbn/rtl/otbn.sv
Comment (ctopal) [reply hw/ip/otbn/rtl/otbn.sv:1202]: If secure wipe does not happen in the conditions you mentioned, we will most likely encounter a status mismatch between model and RTL, so no immediate need for it as far as I understand. Do you mind if we keep mentioned assertion out of scope for this PR?

File: hw/ip/otbn/rtl/otbn.sv
Comment (GregAC) [reply hw/ip/otbn/rtl/otbn.sv:1213]: Remember to change this in the wide reg assertions too

**Review by GregAC — APPROVED** [review]
LGTM modulo some minor assertion comments

**Comment by andreaskurth** [comment]
Here is a patch to fix #14603 (I cannot push to `ctopal:otbn_assertions`):
```patch
From 3ac4cefe0389b98937b9f002edb9b326e18429ab Mon Sep 17 00:00:00 2001
From: Andreas Kurth <adk@lowrisc.org>
Date: Wed, 31 Aug 2022 09:53:22 +0000
Subject: [PATCH] [otbn] Relax assertion for blanking of Adder Y

For `BN.SUBM` with `a >= b` (thus the result of Adder X has the carry
bit set), the result of Adder Y is not used, but it cannot be blanked
solely based on the carry bit (see issue #14603).

This commit relaxes the assertion that checks the blanking of Adder Y to
allow this case, and thereby fixes #14603.

Signed-off-by: Andreas Kurth <adk@lowrisc.org>
---
 hw/ip/otbn/rtl/otbn_alu_bignum.sv | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/hw/ip/otbn/rtl/otbn_alu_bignum.sv b/hw/ip/otbn/rtl/otbn_alu_bignum.sv
index 329897592..d679494ed 100644
--- a/hw/ip/otbn/rtl/otbn_alu_bignum.sv
+++ b/hw/ip/otbn/rtl/otbn_alu_bignum.sv
@@ -858,8 +858,12 @@ module otbn_alu_bignum
           !expected_adder_y_op_shifter_en |-> adder_y_op_shifter_res_blanked == '0,
           clk_i, !rst_ni)

+  // Adder Y must be blanked when its result is not used, with one exception: For `BN.SUBM` with
+  // `a >= b` (thus the result of Adder X has the carry bit set), the result of Adder Y is not used
+  // but it cannot be blanked solely based on the carry bit.
   `ASSERT(BlankingBignumAluYResUsed_A,
-          !adder_y_res_used |-> {x_res_operand_a_mux_out, adder_y_op_b} == '0,
+          !adder_y_res_used && !(operation_i.op == AluOpBignumSubm && adder_x_res[WLEN+1])
+          |-> {x_res_operand_a_mux_out, adder_y_op_b} == '0,
           clk_i, !rst_ni)

   // shifter_res related blanking
```

**Comment by andreaskurth** [comment]
Here is a patch to address @GregAC's feedback:
```patch
From 848e20c96eea4ca75ba0d8581c4bf803bf0e0563 Mon Sep 17 00:00:00 2001
From: Andreas Kurth <adk@lowrisc.org>
Date: Wed, 31 Aug 2022 10:13:37 +0000
Subject: [PATCH] fixup! [otbn] Add assertions for secure wipe

Signed-off-by: Andreas Kurth <adk@lowrisc.org>
---
 hw/ip/otbn/rtl/otbn.sv | 24 ++++++++++++------------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/hw/ip/otbn/rtl/otbn.sv b/hw/ip/otbn/rtl/otbn.sv
index 9188955b1..6bbc46e99 100644
--- a/hw/ip/otbn/rtl/otbn.sv
+++ b/hw/ip/otbn/rtl/otbn.sv
@@ -1205,12 +1205,12 @@ module otbn
       clk_i, !rst_ni)
     // After execution, it's expected to see a change resulting with a nonzero register value
     `ASSERT(SecWipeChangedBaseRegs_A,
-      $rose(busy_secure_wipe) |-> ##[0:$]
-      u_otbn_core.u_otbn_rf_base.gen_rf_base_ff.u_otbn_rf_base_inner.g_rf_flops[i].rf_reg_q !=
-        EccZeroWord &&
-      $changed(
-        u_otbn_core.u_otbn_rf_base.gen_rf_base_ff.u_otbn_rf_base_inner.g_rf_flops[i].rf_reg_q)
-      within $rose(busy_secure_wipe) |-> ##[0:$] $fell(busy_secure_wipe),
+      $rose(busy_secure_wipe) |-> (##[0:$]
+        u_otbn_core.u_otbn_rf_base.gen_rf_base_ff.u_otbn_rf_base_inner.g_rf_flops[i].rf_reg_q !=
+          EccZeroWord &&
+        $changed(
+          u_otbn_core.u_otbn_rf_base.gen_rf_base_ff.u_otbn_rf_base_inner.g_rf_flops[i].rf_reg_q)
+        within $rose(busy_secure_wipe) ##[0:$] $fell(busy_secure_wipe)),
       clk_i, !rst_ni)
   end

@@ -1225,12 +1225,12 @@ module otbn

     // After execution, it's expected to see a change resulting with a nonzero register value
     `ASSERT(SecWipeChangedWideRegs_A,
-            $rose(busy_secure_wipe) |->
-            u_otbn_core.u_otbn_rf_bignum.gen_rf_bignum_ff.u_otbn_rf_bignum_inner.rf[i] !=
-              EccWideZeroWord &&
-            $changed(
-              u_otbn_core.u_otbn_rf_bignum.gen_rf_bignum_ff.u_otbn_rf_bignum_inner.rf[i])
-            within $rose(busy_secure_wipe) |-> ##[0:$] $fell(busy_secure_wipe),
+            $rose(busy_secure_wipe) |-> (##[0:$]
+              u_otbn_core.u_otbn_rf_bignum.gen_rf_bignum_ff.u_otbn_rf_bignum_inner.rf[i] !=
+                EccWideZeroWord &&
+              $changed(
+                u_otbn_core.u_otbn_rf_bignum.gen_rf_bignum_ff.u_otbn_rf_bignum_inner.rf[i])
+              within $rose(busy_secure_wipe) ##[0:$] $fell(busy_secure_wipe)),
           clk_i, !rst_ni)
   end
```

**Comment by andreaskurth** [comment]
CI failed due to #14603, which will be fixed by [this patch](https://github.com/lowRISC/opentitan/pull/14322#issuecomment-1232731399). Once someone with write access to the PR branch (I assume any committer?) has applied the two patches and did `git rebase --autosquash`, this PR should be ready.

**Comment by GregAC** [comment]
> CI failed due to #14603, which will be fixed by [this patch](https://github.com/lowRISC/opentitan/pull/14322#issuecomment-1232731399). Once someone with write access to the PR branch (I assume any committer?) has applied the two patches and did `git rebase --autosquash`, this PR should be ready.

Yes committers should be able to force push to @ctopal's branch. I'll sort this out.

**Comment by GregAC** [comment]
Patches applied and pushed to this PR

**Comment by andreaskurth** [comment]
Looks like some CI checks failed for a reason unrelated to the changes in this PR. Can the failing checks be retried, so we can make sure these changes pass CI?

**Comment by GregAC** [comment]
Had to rebase, one of the CI failures was a known issue we've now fixed related to the Ubuntu 18 -> 20 switch. Should all pass following the rebase.

**Pull request merged**
